[new command] `conan report diff` to inspect diffs between versions and revisions

ErniGH opened the pull request:
Changelog: Feature: New `conan report diff` command to inspect diffs between versions and revisions.
Docs: https://github.com/conan-io/docs/pull/4127

That still need to be done:
* [x] Testing
* [x] Remove the open directives, make this more in line with other Conan commands and let the user open the resulting html file by themselves
* [x] Maybe make a default txt formatter that just spits the diff, and have the html formatter output the necessary files? But the formatters can't create more than 1 file 🤔 
* [x] Have a way to filter what gets diffed - maybe someone can we just care about source code, or only about build-system related changes

ErniGH committed:
wip

ErniGH committed:
wip

Co-Authored-By: Abril Rincón Blanco <5364255+AbrilRBS@users.noreply.github.com>

ErniGH committed:
Merge branch 'erni/new-compare-diff-command' of https://github.com/ErniGH/conan into erni/new-compare-diff-command

AbrilRBS committed:
Fix sidebar overflow behaviour, escape lines of html charcaters

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+    """
+
+    parser.add_argument("-op", "--old-path", help="Path to the old recipe")
+    parser.add_argument("-ov", "--old-reference", help='Old reference "mylib/1.0"')
+    parser.add_argument("-or", "--old-require", help='Old reference "mylib/1.0"')
```
What is the difference? Maybe this can be simplified to just 1?

AbrilRBS replied:
Done, good idea, now the flow is:
* Defining _just_ `--old-reference` will fetch that reference from the remotes
* But if `--old-path` is set, `--old-reference` will be used to set the reference for the conanfile that is being pointed at

The only downside is that if the reference is set in the recipe itself, you'll have to match it on the cmd, but we can work something out if it turns out to be too much of a pain

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+from conan.api.conan_api import ConanAPI
+from conan.api.output import ConanOutput
+from conan.cli.command import conan_command
+from conan.internal.util.files import chdir
```
The idea is that ``cli.commands`` layer doesn't use anything from ``conan.internal``, only from public and mainly from api

AbrilRBS replied:
Yes, this is in fact an unused import leftover from a previous implementation 👍

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+        full_ref, matching_remote = None, None
+        for remote in enabled_remotes:
+            if ref.revision:
+                no_rrev_ref = RecipeReference().loads(reference)
```
```suggestion
                no_rrev_ref = RecipeReference.loads(reference)
```

That is a static method

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+    enabled_remotes = conan_api.remotes.list(args.remote or "*")
+
+    def _download_ref_from_remote(reference):
+        ref = RecipeReference().loads(reference)
```
```suggestion
        ref = RecipeReference.loads(reference)
```

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+        for remote in enabled_remotes:
+            if ref.revision:
+                no_rrev_ref = RecipeReference().loads(reference)
+                no_rrev_ref.revision = None
```
Why removing the revision?

AbrilRBS replied:
The call the `recipe_revisions` below asserts on the revision not being set, so that it can find all revisions for the given reference

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+
+    def _export_recipe_from_path(path_to_conanfile, reference):
+        path = conan_api.local.get_conanfile_path(path_to_conanfile, cwd, py=True)
+        ref = RecipeReference().loads(reference)
```
```suggestion
        ref = RecipeReference.loads(reference)
```

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+    parser.add_argument("-nr", "--new-require", help='New reference "mylib/1.0"')
+
+    parser.add_argument("-s", "--split_diff", action="store_true")
+    parser.add_argument("--encoding", default="utf-8", help="Encoding to read diff")
```
Is the encoding necessary? Conan hardcodes utf-8 for all its generated files

AbrilRBS replied:
We're trying to figure out why git is sometimes outputting non-utf8 characters for some files (My guess is that the file it's diffing itself has a weird encoding/character sequence)

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+    new_export_ref, new_cache_path = _source(args.new_path, args.new_reference, args.new_require)
+
+
+    with open(f"diff{old_export_ref.version}-{new_export_ref.version}.patch", "w") as f:
+        subprocess.run(["git", "diff", old_cache_path, new_cache_path], stdout=f, check=False)
```
Isn't it possible to collect the diffs in memory, and not opening a file? Like ``.run(capture_output=True)``?

AbrilRBS replied:
The current approach (under a TODO to find a better way) is using `check_output_runner` but that's in the `internal` module, and its functionality is not exposed anywhere public. Any suggestions?

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+
+    # Generate small html files
+    output_dir = "html_diffs"
+    subprocess.run(["mkdir", output_dir])
```
Do not use subprocess to create a dir, use ``os.makedirs`` from Python

AbrilRBS committed:
Update conan/cli/commands/compare.py

Co-authored-by: James <memsharded@gmail.com>

AbrilRBS committed:
Update conan/cli/commands/compare.py

Co-authored-by: James <memsharded@gmail.com>

czoido commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+from conan.internal.util.files import chdir
+from conan.api.model import RecipeReference, ListPattern
+
+def output_json(msg):
+    return json.dumps({"greet": msg})
```
I guess this is a leftover? I don't know if a json output would make sense.
```suggestion
```

czoido commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,233 @@
+    return json.dumps({"greet": msg})
+
+
+@conan_command(group="Security", formatters={"json": output_json})
```
For this to be consistent with the rest of the commands this should have a plain text output and a html formatter, instead of writing to a  file directly... I don't know if that's what we want to do or not for this command (I suppose this is a particular case)...

memsharded replied:
It is possible to have a command that doesn't fully align with the formatters, and it always write a file to disk, like the ``conan lock create`` command, which doesn't generate the file via formatters.

czoido replied:
Yes, I mentioned in case we want to consider it, I'm fine with the command generating html files. I was wondering it would be useful to have the diff also in the terminal and being able to select between both text and html?

AbrilRBS replied:
I think it's perfectly reasonable to expect this command to only generate one HTML, we can fix the issue that required to split the files, and then it can be used as any other Conan command (and make the text output usefull if piped to `less`!)

AbrilRBS committed:
Cleanup html diff generation, follow formatter style from Conan commands

AbrilRBS committed:
Merge remote-tracking branch 'ernesto/erni/new-compare-diff-command' into erni/new-compare-diff-command

AbrilRBS committed:
Fix imports

AbrilRBS committed:
Simplify parameter handling, fix ref report for download

AbrilRBS committed:
Add search bar

AbrilRBS committed:
Fix list visuals after search

AbrilRBS committed:
Improve performance for large diffs

AbrilRBS committed:
Add json formatter

---------

Co-authored-by: Ernesto de Gracia Herranz <ernestodgh@jfrog.com>

AbrilRBS committed:
Typo

AbrilRBS committed:
Avoid encoding issues by ignoring them

The implementation right now is a bit ugly, will simplify when it's
ready

---------

Co-authored-by: Ernesto de Gracia Herranz <ernestodgh@jfrog.com>

AbrilRBS committed:
Add some basic tests

---------

Co-authored-by: Ernesto de Gracia Herranz <ernestodgh@jfrog.com>

AbrilRBS commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,134 @@
+    diff = _execute_command(f"git diff --no-index {old_cache_path} {new_cache_path}",
+                            # We ignore the errors because git diff returns 1 if there are differences
+                            ignore_error=True)
+
```
```suggestion

    if args.old_path:
        conan_api.remove.recipe(old_export_ref)
    if args.new_path:
        conan_api.remove.recipe(new_export_ref)
```

So that non-realized revisions do not clobber the cache?

AbrilRBS commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,134 @@
+
+
+# TODO: TESTING FOR ENCODING ISSUES
+def _execute_command(cmd, stderr=None, ignore_error=False):
```
Wondering how we should handle this - We just want to run a git diff and I don't think a copy/paste of an internal function is ok, but we can't use the Git helper either.

Maybe a better idea is to move all/some of this into a new API entrypoint and use the internal version from there?

ErniGH committed:
improve test and use "--- a" instead of "diff --git" (path with spaces)

ErniGH committed:
fix

ErniGH committed:
fix path

AbrilRBS commented on conan/cli/formatters/compare/compare_html.py:
```diff
@@ -0,0 +1,119 @@
+                <div><!--placeholder-->
+                {%- for line in diff_text.splitlines() if not line.startswith("index") -%}
+                    {%- if line.startswith('--- a') %}
+                        {%- set filename = line[5:].strip() %}
```
```suggestion
                        {%- set filename = line[len("--- a"):].strip() %}
```

ErniGH committed:
change html test

ErniGH committed:
wip

AbrilRBS committed:
Make test fail when there are patches

AbrilRBS committed:
Test confgiure_source

AbrilRBS committed:
Test for patches

AbrilRBS committed:
Html test

AbrilRBS committed:
Tests

AbrilRBS committed:
test contents

AbrilRBS committed:
Merge branch 'develop2' into erni/new-compare-diff-command

AbrilRBS commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,168 @@
+from conan.internal.util.files import rmdir
+
+
+def _configure_source(conan_api, conanfile_path, ref, remotes):
```
We need som feedback as to where these 2 methods should live

memsharded commented on conan/cli/commands/compare.py:
```diff
@@ -0,0 +1,168 @@
+from conan.api.conan_api import ConanAPI
+from conan.cli.command import conan_command
+from conan.api.model import RecipeReference
+from conan.internal.conan_app import ConanApp
+from conan.internal.errors import conanfile_exception_formatter
+from conan.internal.graph.graph import CONTEXT_HOST
+from conan.internal.graph.profile_node_definer import initialize_conanfile_profile
+from conan.internal.source import retrieve_exports_sources, config_source
+from conan.internal.util.files import rmdir
```
The command layer shouldn't use anymore any ``conan.internal`` details, so all of this must be hidden in the API.
We need to find:
- Command name
- Location in the API

AbrilRBS replied:
Yes, note that most of these come from the functions that we want to hide anyway, see https://github.com/conan-io/conan/pull/18247#discussion_r2097304308

ErniGH committed:
replace("\\", "/") for windows and use buffer to reorderlines

ErniGH committed:
fix

ErniGH committed:
fix

ErniGH committed:
fix new files, and fix tests

AbrilRBS committed:
Rename

AbrilRBS commented on conan/cli/commands/report.py:
```diff
@@ -0,0 +1,189 @@
+@conan_command(group="Security")
+def report(conan_api: ConanAPI, parser, *args):
+    """
+    Gets information about the recipe and its sources.
```
This needs a better text

AbrilRBS committed:
Break everything

AbrilRBS committed:
Don't show repeated files

AbrilRBS committed:
Simplify generation, share json/html inut data

AbrilRBS committed:
Fix styling

AbrilRBS committed:
More expressive check

AbrilRBS committed:
Move to api? This might get reverted later if we find a better place

AbrilRBS committed:
Improve output

AbrilRBS committed:
Add exclude search

AbrilRBS committed:
Don't take into account paths when filtering

AbrilRBS commented on conan/api/subapi/report.py:
I don't think this should have its own api endpoint, but otherwise I couldn't find a way to hide the internal imports, happy to hear some suggestions

AbrilRBS commented on conan/api/subapi/report.py:
```diff
@@ -0,0 +1,171 @@
+    conanfile.folders.set_base_recipe_metadata(recipe_layout.metadata())
+    config_source(export_source_folder, conanfile, conan_api.config.hook_manager)
+
+def _execute_command(cmd, stderr=None, ignore_error=False):
```
This is mostly the same as `check_runner_output` but ignores encoding errors, maybe we can add a flag to that funcion and use it instead

AbrilRBS commented on conan/cli/formatters/report/diff.py:
```diff
@@ -0,0 +1,99 @@
+from conan.api.output import cli_out_write
+from conan.cli.formatters.report.diff_html import diff_html
+
+def _generate_json(result):
```
The idea is the the html generator uses the json representation as a source for its contents, so that we can just pre-process it once

AbrilRBS commented on conan/cli/formatters/report/diff.py:
```diff
@@ -0,0 +1,99 @@
+def _render_diff(content, template, template_folder, **kwargs):
+    from conan import __version__
+    template = Template(template, autoescape=True)
+    def _safe_filename(filename):
```
This function and the few below it are meant to hide the full path and implementation details of the references

AbrilRBS commented on conan/cli/formatters/report/diff_html.py:
```diff
@@ -0,0 +1,137 @@
+            }
+        </style>
+        <script>
+            let includeSearchQuery = "";
```
@uilianries expressed the need to have an exclusion filter and I agree, it muddies the logic a tiny bit but works good

AbrilRBS reviewed (commented):
Some comments

memsharded commented on conan/cli/formatters/report/__init__.py:
```diff
@@ -0,0 +1 @@
+from .diff import format_diff_html, format_diff_txt, format_diff_json
```
We don't use relative imports, but always absolute ones

AbrilRBS replied:
Note that this comes from the graph formatters using this same approach, should we modify it in all the rest of the formatters?

memsharded replied:
Probably better to be consistent, yes, but not in this PR to avoid increasing the diff

memsharded commented on conan/api/subapi/report.py:
```diff
@@ -0,0 +1,171 @@
+        """
+
+        def _source(path_to_conanfile, reference):
+            if path_to_conanfile is None:
+                export_ref, cache_path = _download_ref_from_remote(self.conan_api, reference, remotes)
```
And what if the reference is already in the cache, and doesn't need to be downloaded?

AbrilRBS replied:
Download ref now first looks in the cache, this will make it behave the same as the rest of the conan commands that first try to resolve in the cache

(but the feature might still be off, need to double check)

memsharded replied:
Maybe just rename the function to something like ``get_ref_from_cache_or_remote()``? Or something shorter, but that doesn't look like that it will always download.

AbrilRBS committed:
Remove duplication, don't try tro download cache refs

AbrilRBS commented on conan/internal/util/files.py:
```diff
@@ -137,9 +137,9 @@ def save_files(path, files, encoding="utf-8"):
 
 
-def load(path, encoding="utf-8"):
+def load(path, encoding="utf-8", ignore_encoding_error=False):
```
Note sure if editing load is aceptable, maybe the check_runner should open the handle by itself if we don't want a new argument

memsharded replied:
Yes, probably better not modify this function, ignoring encoding errors doesn't sound great for other cases (I'd also like to understand where those encoding errors came from)

AbrilRBS commented on conan/api/subapi/report.py:
```diff
@@ -0,0 +1,143 @@
+def _download_ref_from_remote(conan_api, reference, enabled_remotes):
+    ref = RecipeReference.loads(reference)
+    full_ref, matching_remote = None, False
+    # The first remote is None, which means local cache for the list subapi methods
```
This is a bit ugly but avoid repeating the calls below

AbrilRBS committed:
Typo

ErniGH committed:
Merge branch 'develop2' into erni/new-compare-diff-command

ErniGH committed:
fix test

memsharded reviewed (commented):
Looking good, almost there

AbrilRBS committed:
Dont modify load function

AbrilRBS committed:
Rename function

memsharded commented on conan/internal/util/runners.py:
```diff
@@ -79,7 +79,7 @@ def detect_runner(command):
 
 
-def check_output_runner(cmd, stderr=None, ignore_error=False):
+def check_output_runner(cmd, stderr=None, ignore_error=False, ignore_encoding_error=False):
```
Have you tried using the ``conan_run()`` instead, which already has a ``ignore=True`` decoding policy as default?

AbrilRBS replied:
Works nicely, good idea

memsharded reviewed (approved):
Looking good, almost there?

AbrilRBS committed:
Use conan_run

AbrilRBS committed:
Merge branch 'develop2' into erni/new-compare-diff-command

AbrilRBS committed:
Hide git diff stderr

AbrilRBS committed:
Revert check_output_runner

AbrilRBS committed:
Fix issues with paths

czoido commented:
I like this, I think this is a very nice addition and it will be very useful for users. 
I think it's great as it is but for the future I would propose having a more friendly left view of the changes similar of what github does for the PR's. With a tree structure and marking with colors the modified, deleted and new, like this:

![image](https://github.com/user-attachments/assets/3a432bbe-c679-460d-b4bc-6c82b84b5a07)
